### PR TITLE
Hackathon 2405

### PR DESCRIPTION
**Pull request by hammerstefan**
A series of improvements and fixes

### CODE REVIEW

File: pipeline_dash/main.py [hunk 1]
```diff
@@ -59,22 +59,26 @@ def recursive_calculate_status(name: JobName, p: PipelineDict, serial: Optional[
         old_serial = False
         if "server" in p:
             if (
-                serial is not None
+                job_data[name].serial is None and serial is not None
```
Comment (rthill91) [inline pipeline_dash/main.py]: I would consider adding parenthesis around the intended groupings of and/or combinations.

File: pipeline_dash/viz/dash/assets/custom_tabulator.js [hunk 2]
```diff
@@ -33,7 +33,9 @@ window.myNamespace = Object.assign({}, window.myNamespace, {
         },
         statusCellFormat: function (cell, formatterPrams, onRendered) {
             rowData = cell.getRow().getData()
-            cell.getElement().style.backgroundColor = rowData._color
+
+            // cell.getElement().style.backgroundColor = rowData._color
```
Comment (rthill91) [inline pipeline_dash/viz/dash/assets/custom_tabulator.js]: dead code?

File: pipeline_dash/viz/dash/components/graph_col.py [hunk 3]
```diff
@@ -60,4 +71,27 @@ def generate(app: dash.Dash, graph: networkx.DiGraph) -> Tuple[dbc.Col, go.Figur
...
+
+    @app.callback(
+        OperatorOutput(Ids.graph, "figure"),
+        Input(Ids.stores.show_annotations, "data"),
+        State(viz_dash.Ids.stores.session_id, "data"),
+        State(viz_dash.Ids.stores.figure_root, "data"),
+        prevent_initial_call=False,
+    )
+    @logged_callback
+    # @pcprofile
```
Comment (rthill91) [inline pipeline_dash/viz/dash/components/graph_col.py:83]: dead code, maybe intentional? There's several of these, but I'll leave just the one comment

File: pipeline_dash/viz/dash/components/jobs_pipeline_fig.py [hunk 4]
```diff
@@ -1,12 +1,15 @@
+import collections
 import itertools
 import time
 import uuid
 from statistics import median
-from typing import Dict, List, Optional
+from typing import Dict, Optional
```
Comment (rthill91) [inline pipeline_dash/viz/dash/components/jobs_pipeline_fig.py:6]: idk that it matters all that much, but with 3.10 as the minimum you should be able to drop all `Dict`, `List`, etc types and just use `dict`/`list`

File: pipeline_dash/viz/dash/components/jobs_pipeline_fig.py [hunk 5]
```diff
@@ -37,20 +40,30 @@ def scale_font_size(scale: float) -> float:
...
-def size_traces(
-    scale: float, node_trace: go.Scatter, edge_traces: Dict[str, go.Scatter], annotations: List[go.layout.Annotation]
-):
+def size_traces(scale: float, node_trace: go.Scatter, edge_traces: Dict[str, go.Scatter]):
     node_trace.marker.size = max(node_trace.marker.size * scale, 2)
     for et in edge_traces.values():
         et.line.width = max(et.line.width * scale, 0.5)
+
+
+def size_annotations(scale: float, annotations: tuple[go.layout.Annotation, ...]):
```
Comment (rthill91) [inline pipeline_dash/viz/dash/components/jobs_pipeline_fig.py:49]: `tuple` here adds some weight to my previous comment about `Dict` vs `dict`. I'd recommend uniformity regardless of which direction you go

File: pipeline_dash/viz/dash/components/jobs_pipeline_fig.py [hunk 6]
```diff
@@ -59,74 +72,100 @@ def generate_plot_figure(graph: networkx.DiGraph) -> go.Figure:
...
+    #                     dict(label="Label:Off", method="relayout", args=[{"annotations": []}]),
+    #                 ]
+    #             ),
+    #             showactive=False,
+    #             pad=dict(b=0, l=0, r=0, t=0),
+    #             x=0.05,
+    #             y=1,
+    #         )
+    #     ]
+    # )
```
Comment (rthill91) [inline pipeline_dash/viz/dash/components/jobs_pipeline_fig.py]: dead code?

File: pyproject.toml [hunk 7]
```diff
@@ -35,6 +35,12 @@ rich-click = "^1.6.0"
 tenacity = "^8.1.0"
 cerberus = "^1.3.4"
 flatdict = "^4.0.1"
+pytest = "^7.2.0"
+yappi = "^1.4.0"
+line-profiler = "^4.0.2"
+line-profiler-pycharm = "^1.1.0"
+flask-debugtoolbar = "^0.13.1"
```
Comment (rthill91) [inline pyproject.toml:42]: I'm seeing import failures trying to run this. Bumping this to `0.15.1` seems to resolve

```
    import flask_debugtoolbar  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ryan/.cache/pypoetry/virtualenvs/pipeline-dash-WTfjvmy7-py3.11/lib/python3.11/site-packages/flask_debugtoolbar/__init__.py", line 8, in <module>
    from werkzeug.urls import url_quote_plus
ImportError: cannot import name 'url_quote_plus' from 'werkzeug.urls'
```

**Review by rthill91 — CHANGES_REQUESTED** [review]

File: pipeline_dash/viz/dash/components/graph_col.py (hunk 3 above)
Comment (hammerstefan) [reply pipeline_dash/viz/dash/components/graph_col.py:83]: these are semi-intentional to make it easy to turn profiling back on for some critical callbacks. Leaving this for now

File: pipeline_dash/viz/dash/components/jobs_pipeline_fig.py (hunk 4 above)
Comment (hammerstefan) [reply pipeline_dash/viz/dash/components/jobs_pipeline_fig.py:6]: yea, probably a holdover of when I first started this project, on py3.7 :joy:

Comment (hammerstefan) [reply pipeline_dash/viz/dash/components/jobs_pipeline_fig.py:6]: Going to overhaul types in a separate PR

File: pipeline_dash/viz/dash/components/jobs_pipeline_fig.py (hunk 6 above)
Comment (hammerstefan) [reply pipeline_dash/viz/dash/components/jobs_pipeline_fig.py]: Yea, holdover from when I tried to make plotly in-figure menus work. But those menus are utter garbage

**Pull request merged**